File: android/src/main/java/com/apollohg/editor/viewer/ViewerDocument.kt
package com.apollohg.editor.viewer

import com.apollohg.editor.ProseViewerConfiguration
import com.apollohg.editor.ProseViewerError
import com.apollohg.editor.ProseViewerSource
import java.security.MessageDigest
import org.json.JSONObject
import uniffi.editor_core.FfiViewerCompileRequest
import uniffi.editor_core.FfiViewerElement
import uniffi.editor_core.FfiViewerMark
import uniffi.editor_core.FfiViewerSourceKind
import uniffi.editor_core.viewerCompile

/** A typed, width-independent projection of Task 1's immutable compiler stream. */
internal data class ViewerListContext(
    val ordered: Boolean,
    /** Rust u32 list index retained exactly for interaction/accessibility consumers. */
    val index: Long,
    val kind: String?,
    val checked: Boolean,
    val isLast: Boolean,
)

/** Identifies the nearest list item and its first/final renderable leaf. */
internal data class ViewerListItemBoundary(
    val identity: Int,
    val nestingDepth: Int,
    val isFirstRenderableLeaf: Boolean,
    val isFinalRenderableLeaf: Boolean,
)

/**
 * One list-item owner on a leaf's full semantic path, ordered outermost first.
 * Every owner keeps its own marker reservation even when a nested descendant is
 * the nearest list item exposed to interaction/accessibility consumers.
 */
internal data class ViewerListItemAncestor(
    val identity: Int,
    val context: ViewerListContext,
    val nestingDepth: Int,
    val isFirstRenderableLeaf: Boolean,
    val isFinalRenderableLeaf: Boolean,
)

internal sealed interface ViewerInline {
    data class Text(val text: String, val marks: List<FfiViewerMark>) : ViewerInline
    /** Rust u32 document position retained exactly; drawing spans never own it. */
    data class Atom(val nodeType: String, val docPos: Long, val attrsJson: String, val label: String) : ViewerInline
}

internal data class ViewerBlock(
    val nodeType: String,
    val depth: Int,
    val inBlockquote: Boolean,
    val listContext: ViewerListContext?,
    val listItemBoundary: ViewerListItemBoundary?,
    val inlines: List<ViewerInline>,
    val listItemAncestors: List<ViewerListItemAncestor> = emptyList(),
    val outermostListItemIdentity: Int? = null,
    val outermostListItemIsLast: Boolean = false,
)

/** Semantic positions live only in [ViewerInline.Atom], never in Android drawing spans. */
internal data class ViewerDocument(
    val semanticKey: String,
    val blocks: List<ViewerBlock>,
    val isEmpty: Boolean,
    val retainedBytes: Long,
    val trailingEmptyTextBlockCount: Int = 0,
)

internal data class ProseViewerRequest(
    val source: ProseViewerSource,
    val configuration: ProseViewerConfiguration,
    val nativeFontRevision: Long = 0,
    val fontEnvironmentRevision: Long = 0,
    val attachmentRevision: Long = 0,
) {
    val compiledCacheKey: String by lazy {
        sha256(listOf(source.value, configuration.configJson, configuration.imagePolicyJson.orEmpty(), if (configuration.imagesEnabled) "1" else "0", mentionPrefix(configuration.configJson).orEmpty(), source.kind).joinToString("\u001f"))
    }
    val themeDigest: String by lazy { sha256(configuration.themeJson.orEmpty()) }
    /** Semantic publication identity; layout/font revisions deliberately do not enter it. */
    val semanticGenerationIdentity: String by lazy {
        sha256(listOf(
            source.kind,
            source.value,
            configuration.configJson,
            configuration.themeJson.orEmpty(),
            configuration.imagePolicyJson.orEmpty(),
            if (configuration.imagesEnabled) "1" else "0",
            if (configuration.collapsesWhenEmpty) "1" else "0",
            mentionPrefix.orEmpty(),
        ).joinToString("\u001f"))
    }
    /** Immutable layout/cache identity including permitted state-only revisions. */
    val generationIdentity: String by lazy {
        sha256(listOf(semanticGenerationIdentity, attachmentRevision.toString(), nativeFontRevision.toString(), fontEnvironmentRevision.toString()).joinToString("\u001f"))
    }
    val mentionPrefix: String? get() = mentionPrefix(configuration.configJson)
}

internal typealias DocumentCompiler = (ProseViewerRequest) -> ViewerDocument

internal fun compileWithRust(request: ProseViewerRequest): ViewerDocument {
    val result = viewerCompile(
        FfiViewerCompileRequest(
            sourceKind = if (request.source is ProseViewerSource.Html) FfiViewerSourceKind.HTML else FfiViewerSourceKind.JSON,
            source = request.source.value,
            configJson = request.configuration.configJson,
            imagesEnabled = request.configuration.imagesEnabled,
            mentionPrefix = request.mentionPrefix,
        )
    )
    try {
        result.error?.let { throw ProseViewerError.compiler(it.domain, it.code, it.message) }
        val compiled = result.value ?: throw ProseViewerError.compiler("viewer", "MISSING_COMPILED_DOCUMENT", "The compiler returned neither a document nor an error.")
        val semanticKey = compiled.semanticKey()
        if (!semanticKey.matches(Regex("[0-9a-f]{64}"))) {
            throw ProseViewerError.compiler("viewer", "INVALID_SEMANTIC_KEY", "The compiler returned an invalid semantic key.")
        }

        data class Builder(
            val nodeType: String,
            val depth: Int,
            val listContext: ViewerListContext?,
            val listItemIdentity: Int?,
            val listItemContext: ViewerListContext?,
            val inlines: MutableList<ViewerInline> = mutableListOf(),
        )

        val stack = mutableListOf<Builder>()
        val rendered = mutableListOf<ViewerBlock>()
        // A list item's terminal spacing belongs after its own direct leaves,
        // before a child list begins. Descendant leaves are retained only as a
        // fallback for an item whose sole renderable content is nested.
        val directLeavesByListItem = mutableMapOf<Int, MutableList<Int>>()
        val descendantLeavesByListItem = mutableMapOf<Int, MutableList<Int>>()
        val listItemDepths = mutableMapOf<Int, Int>()
        var nextListItemIdentity = 0

        fun nearestListContext(builders: List<Builder>): ViewerListContext? = builders.asReversed().firstNotNullOfOrNull { it.listContext }
        fun listItemAncestors(builders: List<Builder>): List<ViewerListItemAncestor> = builders.mapNotNull { builder ->
            val identity = builder.listItemIdentity ?: return@mapNotNull null
            val context = builder.listItemContext ?: return@mapNotNull null
            ViewerListItemAncestor(identity, context, builder.depth, false, false)
        }
        fun appendLeaf(nodeType: String, depth: Int, inlines: List<ViewerInline>, ancestors: List<Builder>) {
            val itemAncestors = listItemAncestors(ancestors)
            rendered += ViewerBlock(
                nodeType = nodeType,
                depth = depth,
                inBlockquote = ancestors.any { it.nodeType == "blockquote" },
                listContext = nearestListContext(ancestors),
                listItemBoundary = null,
                inlines = inlines,
                listItemAncestors = itemAncestors,
                outermostListItemIdentity = itemAncestors.firstOrNull()?.identity,
                outermostListItemIsLast = itemAncestors.firstOrNull()?.context?.isLast == true,
            )
            itemAncestors.forEach { ancestor ->
                descendantLeavesByListItem.getOrPut(ancestor.identity) { mutableListOf() } += rendered.lastIndex
            }
            itemAncestors.lastOrNull()?.let { nearest ->
                directLeavesByListItem.getOrPut(nearest.identity) { mutableListOf() } += rendered.lastIndex
            }
        }

        compiled.elements().forEach { element ->
            when (element) {
                is FfiViewerElement.BlockStart -> {
                    val context = listContext(element.listContextJson)
                    val identity = if (context != null) nextListItemIdentity++ else null
                    identity?.let { listItemDepths[it] = element.depth.toInt() }
                    stack += Builder(
                        element.nodeType,
                        element.depth.toInt(),
                        context,
                        identity,
                        if (identity == null) null else context,
                    )
                }
                is FfiViewerElement.TextRun -> stack.lastOrNull()?.inlines?.add(ViewerInline.Text(element.text, element.marks))
                is FfiViewerElement.InlineAtom -> stack.lastOrNull()?.inlines?.add(
                    ViewerInline.Atom(element.nodeType, u32(element.docPos), element.attrsJson, element.label)
                )
                is FfiViewerElement.BlockAtom -> appendLeaf(
                    element.nodeType,
                    stack.lastOrNull()?.depth ?: 0,
                    listOf(ViewerInline.Atom(element.nodeType, u32(element.docPos), element.attrsJson, element.label)),
                    stack,
                )
                FfiViewerElement.BlockEnd -> {
                    val builder = stack.removeLastOrNull() ?: return@forEach
                    // Containers are represented by inherited context. Every text block,
                    // including an empty paragraph, remains a leaf for list boundaries.
                    if (builder.nodeType !in CONTAINER_BLOCKS && builder.listItemIdentity == null) {
                        appendLeaf(builder.nodeType, builder.depth, builder.inlines, stack + builder)
                    }
                }
            }
        }
        descendantLeavesByListItem.forEach { (identity, descendantLeaves) ->
            val leaves = directLeavesByListItem[identity]?.takeIf { it.isNotEmpty() } ?: descendantLeaves
            val first = leaves.firstOrNull() ?: return@forEach
            val final = leaves.last()
            leaves.forEach { index ->
                val updatedAncestors = rendered[index].listItemAncestors.map { ancestor ->
                    if (ancestor.identity == identity) {
                        ancestor.copy(
                            nestingDepth = listItemDepths[identity] ?: ancestor.nestingDepth,
                            isFirstRenderableLeaf = index == first,
                            isFinalRenderableLeaf = index == final,
                        )
                    } else ancestor
                }
                val nearest = updatedAncestors.lastOrNull()
                rendered[index] = rendered[index].copy(
                    listItemBoundary = nearest?.let {
                        ViewerListItemBoundary(it.identity, it.nestingDepth, it.isFirstRenderableLeaf, it.isFinalRenderableLeaf)
                    },
                    listItemAncestors = updatedAncestors,
                )
            }
        }
        val fallback = if (rendered.isEmpty() && !compiled.isEmpty()) listOf(ViewerBlock("paragraph", 0, false, null, null, emptyList())) else rendered
        val admittedAttachmentCount = fallback.count { block ->
            block.nodeType == "image" && ViewerImageAttachment.sourceAndDeclaredSize(block) != null
        }
        if (admittedAttachmentCount > ViewerImageAttachment.MAXIMUM_ADMITTED_ATTACHMENTS) {
            throw ProseViewerError.compiler(
                "viewer",
                "ATTACHMENT_LIMIT_EXCEEDED",
                "The document exceeds the maximum admitted image attachment count.",
            )
        }
        return ViewerDocument(
            semanticKey = semanticKey,
            blocks = fallback,
            isEmpty = compiled.isEmpty(),
            retainedBytes = compiled.retainedBytesDecimal().toLongOrNull() ?: 0,
            trailingEmptyTextBlockCount = compiled.trailingEmptyTextBlockCount().toInt(),
        )
    } finally {
        result.destroy()
    }
}

private val CONTAINER_BLOCKS = setOf("doc", "blockquote", "bulletList", "orderedList", "taskList", "listItem")

internal fun listContext(json: String?): ViewerListContext? = runCatching {
    json ?: return@runCatching null
    val value = JSONObject(json)
    val index = if (value.has("index")) u32(value.opt("index")) else 1L
    ViewerListContext(
        value.optBoolean("ordered"),
        index,
        value.optionalString("kind"),
        value.optBoolean("checked"),
        value.optBoolean("isLast"),
    )
}.getOrNull()

/**
 * `optString(key, null)` cannot express an absent value: its fallback is
 * declared non-null, and a present JSON null coerces to the string "null".
 */
internal fun JSONObject.optionalString(key: String): String? =
    if (isNull(key)) null else optString(key)

/** JSON and UniFFI may expose Rust u32 values through different Kotlin number types. */
private fun u32(value: Any?): Long {
    val parsed = value?.toString()?.toLongOrNull()
        ?: throw IllegalArgumentException("Expected an unsigned 32-bit semantic value.")
    require(parsed in 0L..0xFFFF_FFFFL) { "Semantic value is outside Rust u32 range." }
    return parsed
}

private fun mentionPrefix(configJson: String): String? = runCatching {
    val root = JSONObject(configJson)
    root.optJSONObject("mentions")?.optionalString("prefix") ?: root.optionalString("mentionPrefix")
}.getOrNull()

internal fun sha256(value: String): String = MessageDigest.getInstance("SHA-256").digest(value.toByteArray(Charsets.UTF_8)).joinToString("") { "%02x".format(it) }

File: ios/Viewer/ViewerDocument.swift
import CryptoKit
import Foundation

/// The source accepted by both the UIKit and Fabric viewer surfaces.
public enum ProseViewerSource: Hashable {
    case json(String)
    case html(String)

    var kind: FfiViewerSourceKind {
        switch self {
        case .json: .json
        case .html: .html
        }
    }

    var value: String {
        switch self {
        case let .json(value), let .html(value): value
        }
    }
}

/// Immutable configuration for one viewer generation.
public struct ProseViewerConfiguration: Hashable {
    public var configJSON: String
    public var themeJSON: String?
    public var imagePolicyJSON: String?
    public var imagesEnabled: Bool
    public var collapsesWhenEmpty: Bool

    public init(
        configJSON: String = "{}",
        themeJSON: String? = nil,
        imagePolicyJSON: String? = nil,
        imagesEnabled: Bool = true,
        collapsesWhenEmpty: Bool = true
    ) {
        self.configJSON = configJSON
        self.themeJSON = themeJSON
        self.imagePolicyJSON = imagePolicyJSON
        self.imagesEnabled = imagesEnabled
        self.collapsesWhenEmpty = collapsesWhenEmpty
    }
}

struct ProseViewerRequest: Hashable {
    let source: ProseViewerSource
    let configuration: ProseViewerConfiguration
    let nativeFontRevision: UInt64
    let fontEnvironmentRevision: UInt64
    let attachmentRevision: UInt64

    init(
        source: ProseViewerSource,
        configuration: ProseViewerConfiguration,
        nativeFontRevision: UInt64 = 0,
        fontEnvironmentRevision: UInt64 = 0,
        attachmentRevision: UInt64 = 0
    ) {
        self.source = source
        self.configuration = configuration
        self.nativeFontRevision = nativeFontRevision
        self.fontEnvironmentRevision = fontEnvironmentRevision
        self.attachmentRevision = attachmentRevision
    }

    var compiledCacheKey: String {
        let mentionPrefix = Self.mentionPrefix(in: configuration.configJSON) ?? ""
        let input = [
            source.value,
            configuration.configJSON,
            configuration.imagePolicyJSON ?? "",
            configuration.imagesEnabled ? "1" : "0",
            mentionPrefix,
            source.kind == .json ? "json" : "html",
        ].joined(separator: "\u{1F}")
        return SHA256Digest.hex(input)
    }

    var themeDigest: String { SHA256Digest.hex(configuration.themeJSON ?? "") }

    /// Includes every input that makes a mounted generation genuinely different.
    var generationIdentity: String {
        SHA256Digest.hex([
            compiledCacheKey,
            themeDigest,
            configuration.collapsesWhenEmpty ? "1" : "0",
            String(attachmentRevision),
            String(nativeFontRevision),
            String(fontEnvironmentRevision),
        ].joined(separator: "\u{1F}"))
    }

    var mentionPrefix: String? { Self.mentionPrefix(in: configuration.configJSON) }

    private static func mentionPrefix(in json: String) -> String? {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return ((root["mentions"] as? [String: Any])?["prefix"] as? String)
            ?? (root["mentionPrefix"] as? String)
    }
}

public enum ProseViewerError: Error, Equatable {
    case compiler(domain: String, code: String, message: String)
    case hostContract(message: String)
    case layout(message: String)

    var domain: String {
        switch self {
        case let .compiler(domain, _, _): domain
        case .hostContract: "viewer.host"
        case .layout: "viewer.layout"
        }
    }

    var code: String {
        switch self {
        case let .compiler(_, code, _): code
        case .hostContract: "INVALID_WIDTH"
        case .layout: "LAYOUT_FAILED"
        }
    }

    var message: String {
        switch self {
        case let .compiler(_, _, message), let .hostContract(message), let .layout(message): message
        }
    }
}

struct ViewerParagraph: Hashable {
    let text: String
}

/// Width-independent native projection of the Rust compiled document.
struct ViewerDocument: Hashable {
    let semanticKey: String
    let paragraphs: [ViewerParagraph]
    let isEmpty: Bool
    let retainedBytes: Int

    init(semanticKey: String, paragraphs: [ViewerParagraph], isEmpty: Bool, retainedBytes: Int) {
        self.semanticKey = semanticKey
        self.paragraphs = paragraphs
        self.isEmpty = isEmpty
        self.retainedBytes = retainedBytes
    }

    init(compiled: ViewerCompiledDocument) throws {
        semanticKey = compiled.semanticKey()
        isEmpty = compiled.isEmpty()
        retainedBytes = Int(compiled.retainedBytesDecimal()) ?? 0

        var paragraphs: [ViewerParagraph] = []
        var text = ""
        var isInBlock = false
        for element in compiled.elements() {
            switch element {
            case .blockStart:
                if isInBlock { paragraphs.append(ViewerParagraph(text: text)); text = "" }
                isInBlock = true
            case let .textRun(text: value, marks: _):
                text.append(value)
            case let .inlineAtom(nodeType: _, docPos: _, attrsJson: _, label: label),
                 let .blockAtom(nodeType: _, docPos: _, attrsJson: _, label: label):
                text.append(label)
            case .blockEnd:
                paragraphs.append(ViewerParagraph(text: text))
                text = ""
                isInBlock = false
            }
        }
        if isInBlock || (!text.isEmpty && paragraphs.isEmpty) {
            paragraphs.append(ViewerParagraph(text: text))
        }
        self.paragraphs = paragraphs.isEmpty ? [ViewerParagraph(text: "")] : paragraphs
    }
}

private enum SHA256Digest {
    static func hex(_ value: String) -> String {
        let data = Data(value.utf8)
        return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}
